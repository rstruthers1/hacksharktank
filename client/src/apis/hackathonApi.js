// hackathonApi.js
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

const baseURL = process.env.REACT_APP_API_URL || "";

export const hackathonApi = createApi({
    reducerPath: 'hackathonApi',
    baseQuery: fetchBaseQuery(
        {baseUrl: `${baseURL}/`}), // Set your base URL here
    endpoints: (builder) => ({
        createHackathon: builder.mutation({
            query: (hackathonData) => ({
                url: 'hackathons', // Your endpoint path
                method: 'POST',
                body: hackathonData,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
        }),
        updateHackathon: builder.mutation({
            query: (hackathonData) => ({
                url: `hackathons/${hackathonData.id}`, // Your endpoint path
                method: 'PUT',
                body: hackathonData,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'Hackathons'},
                { type: 'Hackathon', id: arg.id }],
        }),
        getHackathons: builder.query({
            query: () => ({
                url: 'hackathons', // Your endpoint path
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'Hackathons'}],
        }),
        getHackathon: builder.query({
            query: (hackathonId) => ({
                url: `hackathons/${hackathonId}`, // Your endpoint path
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'Hackathon', id: arg }],
        }),
        createHackathonUserRole: builder.mutation({
            query: (hackathonUserRoleData) => ({
                url: 'hackathons/users/roles', // Your endpoint path
                method: 'POST',
                body: hackathonUserRoleData,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg.hackathonId }],
        }),
        deleteHackathonUserRole: builder.mutation({
            query: (hackathonUserRoleData) => ({
                url: `hackathons/${hackathonUserRoleData.hackathonId}/users/${hackathonUserRoleData.userId}/roles/${hackathonUserRoleData.roleName}`, // Your endpoint path
                method: 'DELETE',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg.hackathonId }],
        }),
        getHackathonUsers: builder.query({
            query: (hackathonId) => ({
                url: `hackathons/${hackathonId}/users`, // Your endpoint path
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg }],
        }),
        deleteHackathonUser: builder.mutation({
            query: (hackathonUserData) => ({
                url: `hackathons/${hackathonUserData.hackathonId}/users/${hackathonUserData.userId}`, // Your endpoint path
                method: 'DELETE',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonUsers', id: arg.hackathonId }],
        }),
        // You can add more endpoints here
    }),
});

export const {
    useCreateHackathonMutation,
    useUpdateHackathonMutation,
    useGetHackathonsQuery,
    useCreateHackathonUserRoleMutation,
    useGetHackathonQuery,
    useGetHackathonUsersQuery,
    useDeleteHackathonUserRoleMutation,
    useDeleteHackathonUserMutation,
} = hackathonApi;
