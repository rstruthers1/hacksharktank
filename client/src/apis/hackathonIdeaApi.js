import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

const baseURL = process.env.REACT_APP_API_URL || "";

export const hackathonIdeaApi = createApi({
    reducerPath: 'hackathonIdeaApi',
    baseQuery: fetchBaseQuery(
        {baseUrl: `${baseURL}/`}), // Set your base URL here
    endpoints: (builder) => ({
        getHackathonIdeas: builder.query({
            query: (hackathonId) => ({
                url: `hackathons/${hackathonId}/ideas`,
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
        }),
       createHackathonIdea: builder.mutation({
                query: (hackathonIdeaData) => ({
                    url: `hackathons/${hackathonIdeaData.hackathonId}/ideas`,
                    method: 'POST',
                    body: hackathonIdeaData.data,
                    // Add a JWT token to the request headers if the user is logged in
                    headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
                }),
                invalidatesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
            }),
        updateHackathonIdea: builder.mutation({
            query: (hackathonIdeaData) => ({
                url: `hackathons/${hackathonIdeaData.hackathonId}/ideas`,
                method: 'PUT',
                body: hackathonIdeaData.data,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
        }),
        deleteHackathonIdea: builder.mutation({
            query: ({hackathonId, ideaId}) => ({
                url: `hackathons/${hackathonId}/ideas/${ideaId}`,
                method: 'DELETE',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'HackathonIdeas'}],
        }),
    })
});

export const {
    useGetHackathonIdeasQuery,
    useCreateHackathonIdeaMutation,
    useDeleteHackathonIdeaMutation,
    useUpdateHackathonIdeaMutation
} = hackathonIdeaApi;
