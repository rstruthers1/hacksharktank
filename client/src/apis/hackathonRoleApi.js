import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';

const baseURL = process.env.REACT_APP_API_URL || "";

export const hackathonRoleApi = createApi({
  reducerPath: 'hackathonRoleApi',
    baseQuery: fetchBaseQuery(
        {baseUrl: `${baseURL}/`}), // Set your base URL here
    endpoints: (builder) => ({
      getHackathonRoles: builder.query({
        query: () => ({
          url: `hackathon-roles`,
          method: 'GET',
          // Add a JWT token to the request headers if the user is logged in
          headers: {'Authorization': `JWT ${localStorage.getItem('token')}`},
        }),
        providesTags: (result, error, arg) => [{type: 'HackathonRoles', id: arg}],
      }),
    })

});

export const {useGetHackathonRolesQuery} = hackathonRoleApi;