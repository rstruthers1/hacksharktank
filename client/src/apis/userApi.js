// userApi.js
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

const baseURL = process.env.REACT_APP_API_URL || "";

export const userApi = createApi({
    reducerPath: 'userApi',
    baseQuery: fetchBaseQuery(
        { baseUrl: `${baseURL}/` }), // Set your base URL here
    endpoints: (builder) => ({
        registerUser: builder.mutation({
            query: (userData) => ({
                url: 'users',
                method: 'POST',
                body: userData,
            }),
        }),
        loginUser: builder.mutation({
            query: (userData) => ({
                url: 'users/login',
                method: 'POST',
                body: userData,
            }),
        }),
        searchUsers: builder.query({
            query: (searchTerm) => ({
                url: `users/search?searchTerm=${searchTerm}`,
                method: 'GET',
            }),
        }),
        getUser: builder.query({
            query: (userId) => ({
                url: `users/${userId}`,
                method: 'GET',
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            providesTags: (result, error, arg) => [{ type: 'User', id: arg }],
        }),
        updateUser: builder.mutation({
            query: (userData) => ({
                url: `users/${userData.id}`,
                method: 'PUT',
                body: userData,
                // Add a JWT token to the request headers if the user is logged in
                headers: { 'Authorization': `JWT ${localStorage.getItem('token')}` },
            }),
            invalidatesTags: (result, error, arg) => [{ type: 'User', id: arg.id }],
        }),
        // You can add more endpoints here
    }),
});

// Export the auto-generated hook for the `registerUser` and `loginUser` mutation
export const {
    useRegisterUserMutation,
    useLoginUserMutation,
    useLazySearchUsersQuery,
    useGetUserQuery,
    useUpdateUserMutation
} = userApi;
