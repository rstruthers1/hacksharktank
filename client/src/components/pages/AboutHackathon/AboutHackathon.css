.specific-hackathon {
    font-family: Arial, sans-serif;
    line-height: 1.6;
    max-width: 100%;
    margin: auto;
    padding: 20px;
}

.specific-hackathon h1, .specific-hackathon h2 {
    color: #333;
}

.specific-hackathon section {
    margin-bottom: 20px;
}

.register-btn {
    background-color: #28a745;
    color: white;
    border: none;
    padding: 10px 20px;
    border-radius: 5px;
    cursor: pointer;
}

.register-btn:hover {
    background-color: #218838;
}
