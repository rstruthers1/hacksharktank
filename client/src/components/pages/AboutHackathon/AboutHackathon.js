import React from 'react';
import './AboutHackathon.css';
import {useParams} from "react-router-dom";
import {formatDateTimeAsDate} from "../../../utils/dateTimeUtils";
import {useGetHackathonQuery} from "../../../apis/hackathonApi";

const AboutHackathon = () => {
    const {hackathonId} = useParams();
    const {
        data: hackathon,
        error: getHackathonError,
        isLoading: getHackathonIsLoading
    } = useGetHackathonQuery(hackathonId);

    return (
        <div className="specific-hackathon">
            {getHackathonError ? (
                <>Oh no, there was an error</>
            ) : getHackathonIsLoading ? (
                <>Loading...</>
            ) : hackathon && (
                <>
                    <h1>{hackathon.eventName}</h1>
                <p>{hackathon.description}</p>

                <section className="event-details">
                <h2>Event Details</h2>
                <p>Start Date: {formatDateTimeAsDate(hackathon.startDate)}</p>
                <p>End Date: {formatDateTimeAsDate(hackathon.endDate)}</p>
                <p>Location: Virtual</p>

                </section>
                    </>
                )

            }
        </div>
    );
};

export default AboutHackathon;
