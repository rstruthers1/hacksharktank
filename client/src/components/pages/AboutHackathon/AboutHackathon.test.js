import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { useParams } from 'react-router-dom';
import { useGetHackathonQuery } from '../../../apis/hackathonApi';
import AboutHackathon from './AboutHackathon';
import '@testing-library/jest-dom';
import renderer from "react-test-renderer";

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: jest.fn(),
}));

jest.mock('../../../apis/hackathonApi', () => ({
    useGetHackathonQuery: jest.fn(),
}));

describe('AboutHackathon', () => {
    it('should render correctly', () => {
        useParams.mockReturnValue({ hackathonId: '1' });
        useGetHackathonQuery.mockReturnValue({
            data: null,
            error: null,
            isLoading: true,
        });
        const tree = renderer.create(<AboutHackathon/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly', async () => {
        useParams.mockReturnValue({ hackathonId: '1' });
        useGetHackathonQuery.mockReturnValue({
            data: {
                eventName: 'Test Hackathon',
                description: 'This is a test hackathon',
                startDate: new Date(),
                endDate: new Date(),
            },
            error: null,
            isLoading: false,
        });

        const { getByText } = render(<AboutHackathon />);

        await waitFor(() => {
            expect(getByText('Test Hackathon')).toBeInTheDocument();
            expect(getByText('This is a test hackathon')).toBeInTheDocument();
        });
    });

    it('displays loading state', () => {
        useParams.mockReturnValue({ hackathonId: '1' });
        useGetHackathonQuery.mockReturnValue({
            data: null,
            error: null,
            isLoading: true,
        });

        const { getByText } = render(<AboutHackathon />);

        expect(getByText('Loading...')).toBeInTheDocument();
    });

    it('displays error state', () => {
        useParams.mockReturnValue({ hackathonId: '1' });
        useGetHackathonQuery.mockReturnValue({
            data: null,
            error: 'Error',
            isLoading: false,
        });

        const { getByText } = render(<AboutHackathon />);

        expect(getByText('Oh no, there was an error')).toBeInTheDocument();
    });
});