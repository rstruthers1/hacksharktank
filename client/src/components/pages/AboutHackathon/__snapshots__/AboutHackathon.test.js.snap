// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutHackathon should render correctly 1`] = `
<div
  className="specific-hackathon"
>
  Loading...
</div>
`;
